api/sendfild/service/service_test.go

```diff
@@ -0,0 +1,646 @@
+}
+
+func TestListTxnsUpdatedBefore(t *testing.T) {
+	c, lc, dAddr, cleanup := requireSetup(t, ctx)
+	defer cleanup()
+	addr1 := requireLotusAddress(t, ctx, lc)
+	requireSendFil(t, ctx, c, dAddr.String(), addr1.String(), oneFil, false)
+	requireSendFil(t, ctx, c, dAddr.String(), addr1.String(), oneFil, false)
+	t3 := requireSendFil(t, ctx, c, dAddr.String(), addr1.String(), oneFil, false)
+	res, err := c.ListTxns(ctx, &pb.ListTxnsRequest{UpdatedBefore: t3.UpdatedAt})
+	require.NoError(t, err)
+	require.Len(t, res.Txns, 2)
+}
+
+func TestListTxnsUpdatedAfterBefore(t *testing.T) {
+	c, lc, dAddr, cleanup := requireSetup(t, ctx)
+	defer cleanup()
+	addr1 := requireLotusAddress(t, ctx, lc)
+	t1 := requireSendFil(t, ctx, c, dAddr.String(), addr1.String(), oneFil, false)
+	requireSendFil(t, ctx, c, dAddr.String(), addr1.String(), oneFil, false)
+	t3 := requireSendFil(t, ctx, c, dAddr.String(), addr1.String(), oneFil, false)
+	res, err := c.ListTxns(ctx, &pb.ListTxnsRequest{UpdatedAfter: t1.UpdatedAt, UpdatedBefore: t3.UpdatedAt})
+	require.NoError(t, err)
+	require.Len(t, res.Txns, 1)
+}
+
+func TestListTxnsOrder(t *testing.T) {
+	c, lc, dAddr, cleanup := requireSetup(t, ctx)
+	defer cleanup()
+	addr1 := requireLotusAddress(t, ctx, lc)
+	requireSendFil(t, ctx, c, dAddr.String(), addr1.String(), oneFil, false)
+	requireSendFil(t, ctx, c, dAddr.String(), addr1.String(), oneFil, false)
+	requireSendFil(t, ctx, c, dAddr.String(), addr1.String(), oneFil, false)
+	res, err := c.ListTxns(ctx, &pb.ListTxnsRequest{Ascending: false})
+	require.NoError(t, err)
+	require.Len(t, res.Txns, 3)
+	requireTxnsOrder(t, res.Txns, false)
+	res, err = c.ListTxns(ctx, &pb.ListTxnsRequest{Ascending: true})
+	require.NoError(t, err)
+	require.Len(t, res.Txns, 3)
+	requireTxnsOrder(t, res.Txns, true)
+}
+
+func TestListTxnsPaging(t *testing.T) {
+	c, lc, dAddr, cleanup := requireSetup(t, ctx)
+	defer cleanup()
+	addr1 := requireLotusAddress(t, ctx, lc)
+	txFirst := requireSendFil(t, ctx, c, dAddr.String(), addr1.String(), oneFil, false)
+	requireSendFil(t, ctx, c, dAddr.String(), addr1.String(), oneFil, false)
+	requireSendFil(t, ctx, c, dAddr.String(), addr1.String(), oneFil, false)
+	requireSendFil(t, ctx, c, dAddr.String(), addr1.String(), oneFil, false)
+	requireSendFil(t, ctx, c, dAddr.String(), addr1.String(), oneFil, false)
+	requireSendFil(t, ctx, c, dAddr.String(), addr1.String(), oneFil, false)
+	requireSendFil(t, ctx, c, dAddr.String(), addr1.String(), oneFil, false)
+	requireSendFil(t, ctx, c, dAddr.String(), addr1.String(), oneFil, false)
+	requireSendFil(t, ctx, c, dAddr.String(), addr1.String(), oneFil, false)
+	txLast := requireSendFil(t, ctx, c, dAddr.String(), addr1.String(), oneFil, false)
+
+	pageResults := func(ascending bool) {
+		numPages := 0
+		more := true
+		var moreToken int64 = 0
+		for more {
+			req := &pb.ListTxnsRequest{CreatedAfter: txFirst.CreatedAt, CreatedBefore: txLast.CreatedAt, Limit: 3, Ascending: ascending}
```

Comment:
This paging test makes sure that paging works properly when `CreatedAfter` and `CreatedBefore` constraints are also provided, both ascending and descending. This was tricky because paging also internally uses a `created_at` filter via the `MoreToken`... so it works!